### PR TITLE
(Chore) Refactor Generic Report Feature

### PR DESCRIPTION
**Pull request by sfount**
This pull request updates the generic report feature with comments and code review feedback as well as simplifying/ cleaning up the controllers. 

* Address console log and format issues as reviewed by @mbayopanda https://github.com/IMA-WorldHealth/bhima-2.X/pull/832
* Options passed to report configuration modal instead of hardcoded in the controller 
* Moves all Modal and $http interaction into separate service and out of controller 
* Handles edge cases on the server

---

 - [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
 - [x] Run integration tests.
 - [x] Run end-to-end tests.
 - [x] Accurately described the changes your are making in this pull request.

### CODE REVIEW

File: client/src/partials/reports/modals/accounts_chart.config.js [hunk 1]
```diff
@@ -1,50 +1,49 @@
...
         $state.reload();
        })
       .catch(function (error) {
+        var INTERNAL_SERVER_ERROR = 500;
         vm.$loading = false;
+
+        if (error.status === INTERNAL_SERVER_ERROR) {
+          ModalInstance.dismiss();
+        }
+
```
Comment (mbayopanda) [inline client/src/partials/reports/modals/accounts_chart.config.js:41]: 👍

File: client/src/partials/reports/reports.js [hunk 2]
```diff
@@ -48,20 +49,8 @@ function ReportsController($state, SavedReports, Modal) {
...
-    // template : /partials/reports/modals/:report_key:.modal.html
-    // controller : :report_key:Controller
-    var templateString = '/partials/reports/modals/'.concat(keyTarget, '.modal.html');
-    var controllerString = keyTarget.concat('Controller as ReportConfigCtrl');
 
-    return Modal.open({
-      templateUrl : templateString,
-      controller : controllerString,
-      size : 'md'
-    });
```
Comment (mbayopanda) [inline client/src/partials/reports/reports.js:65]: 👍

File: server/lib/ReportManager.js [hunk 3]
```diff
@@ -97,6 +98,13 @@ class ReportManager {
       throw new BadRequest(`The application does not support rendering ${options.renderer}.`, 'ERRORS.INVALID_RENDERER');
     }
 
+    // @TODO user information could be determined by report manager, removing the need for this check
+    if (this.options.saveReport && !this.options.user) {
+      let invalidSaveDescription = 'Report cannot be saved without providing a `user` entity to ReportManager';
+      console.error(invalidSaveDescription);
+      throw new InternalServerError(invalidSaveDescription);
```
Comment (mbayopanda) [inline server/lib/ReportManager.js:105]: The `console.error` is needed. The question is if we can use `console.error` in combination with `throw new ...`

**Review by mbayopanda — APPROVED** [review]
👍 LGTM

File: server/lib/ReportManager.js (hunk 3 above)
Comment (jniles) [reply server/lib/ReportManager.js:105]: This looks fine to me in terms of code.

I would encourage that instead of `console`, we use `winston` so that we can control the type and destination of all logged text.  It doesn't need to be done in the PR, as the whole server will need to be reviewed so that errors are properly logged, and useful information is logged.

File: server/controllers/report.js [hunk 4]
```diff
@@ -38,9 +61,17 @@ function list(req, res, next) {
     .done();
 }
 
+/**
+ * @function sendArchived
```
Comment (jniles) [inline server/controllers/report.js:65]: Nice comments :+1:

File: client/src/partials/reports/modals/accounts_chart.config.js [hunk 5]
```diff
@@ -1,50 +1,49 @@
...
 
-    $http.get(url, { params : pdfParams })
+    SavedReports.requestPDF(url, report, options)
       .then(function (result) {
         vm.$loading = false;
         ModalInstance.dismiss();
         $state.reload();
        })
       .catch(function (error) {
+        var INTERNAL_SERVER_ERROR = 500;
```
Comment (jniles) [inline client/src/partials/reports/modals/accounts_chart.config.js:35]: In the future, I think this would be better in `bhConstants`.

**Review by jniles — APPROVED** [review]
All these changes look good to me.  Very nice and clean, and the comments provide valuable context.
